File: src/api/image.ts
import { call, listen, type UnlistenFn } from './_core';
import type { SDStatus, ImageGenerationResult, SDDownloadProgress, SDGenerationProgress } from '../types';

export const imageApi = {
  getSDStatus: (): Promise<SDStatus> => call<SDStatus>('get_sd_status'),
  openSDFolder: (): Promise<void> => call<void>('open_sd_folder'),

  downloadSD: (): Promise<void> => call<void>('download_sd'),

  generateImage: (
    prompt: string,
    negativePrompt?: string,
    width?: number,
    height?: number,
    steps?: number,
    seed?: number,
  ): Promise<ImageGenerationResult> =>
    call<ImageGenerationResult>('generate_image_sd', {
      prompt,
      negativePrompt,
      width,
      height,
      steps,
      seed,
    }),

  onDownloadProgress: (cb: (progress: SDDownloadProgress) => void): Promise<UnlistenFn> =>
    listen<SDDownloadProgress>('sd-download-progress', (e) => {
      if (e.payload) cb(e.payload);
    }),

  onGenerationProgress: (cb: (data: SDGenerationProgress) => void): Promise<UnlistenFn> =>
    listen<SDGenerationProgress>('sd-generation-progress', (e) => {
      if (e.payload) cb(e.payload);
    }),
};

File: src-tauri/src/commands/image.rs
use crate::sd_engine::{ImageGenerationResult, SDEngine, SDStatus};

#[tauri::command]
pub fn get_sd_status() -> SDStatus {
    SDEngine::get_status()
}

#[tauri::command]
pub async fn download_sd(window: tauri::Window) -> Result<(), String> {
    SDEngine::download_all(window).await
}

#[tauri::command]
pub async fn generate_image_sd(
    prompt: String,
    negative_prompt: Option<String>,
    width: Option<u32>,
    height: Option<u32>,
    steps: Option<u32>,
    seed: Option<i64>,
    window: tauri::Window,
) -> Result<ImageGenerationResult, String> {
    SDEngine::generate_image(
        prompt,
        negative_prompt,
        width.unwrap_or(512),
        height.unwrap_or(512),
        steps.unwrap_or(15),
        seed,
        window,
    )
    .await
}
